design-proposals/klip-63-pause-resume-persistent-queries.md

```diff
@@ -0,0 +1,66 @@
+# KLIP-63: PAUSE and RESUME for persistent queries 
+
+**Author**: Jim Hughes (jnh5y) | 
+**Release Target**: Unknown | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** Adds the ability to pause the processing of a persistent query.
+           
+## Motivation and background
+
+Users may wish to pause a query for a number of reasons and then resume processing.  As examples, users may wish to modify data pipelines built from persistent queries or to manage computational resources.
+
+Presently to manage processing, they would have to terminate a query and recreate it.  This approach means that any state would have to be restored; the goal of this work is to add the capability to pause processing temporarily while allowing Kafka consumers and state stores to be retained.
```

Comment:
How is this affecting data producers? Is the assumption here that we will be able to hold/buffer data upstream as long as the query is paused? What happens if the query is paused indefinitely? Is there going to be a threshold?

Reply:
I believe the output from a KafkaStreams topology would be sent out to a Kafka producer during the execution of the topology. 

Pausing a query will not interrupt a single execution of the topology, so the output from a topology should be written out to the Kafka producers a-ok.  

If a query is paused indefinitely, the query is paused.  At the moment, there is no threshold or timeout.

Reply:
If a query is pause for too long, and the input topic is subject to data purging via retention, it would be an issue semantically, but it should not crash the query.

The consumer would try to fetch from a non-existing offset when resumed, and the broker would return an error resulting in an exception. Kafka Streams would catch this exception and would reset the offset based on `auto.offset.reset` configuration.

Reply:
> This approach means that any state would have to be restored

This might be miss-leading -- the state is lost and the changelog topic are deleted if you terminate a query. Thus, you cannot _restore_ state -- if (but only if) you still have all your input data, you can only _recompute_ it though.

Reply:
To the first comment, is that situation worth calling out in the KLIP?

For the second, nice call, I am replacing `restored` with `recomputed`.

Reply:
I think it's not necessary to discuss on the KLIP. Would go down the rabbit whole too much -- just wanted to explain the situation (in case anybody is interested in the technical details).